contracts/simple-dex/lib.rs

```diff
@@ -0,0 +1,454 @@
+#![cfg_attr(not(feature = "std"), no_std, no_main)]
+
+/// SimpleDex contract
+/// Basically a Balancer v1 pool with all weights equal
+///
+/// Supports swaps and all assets withdrawals / deposits
+/// Keeps track of LP tokens in a map
+///
+/// - single asset deposit / withdrawal is left as an exerciese
+/// - so is implementing LP shares as a token in PSP22 stadard (mintable & burnable)
+#[ink::contract]
+mod dex {
+
+    use ink::{
+        codegen::EmitEvent, prelude::vec::Vec, reflect::ContractEventBase, storage::Mapping,
+    };
+    use psp22_traits::{PSP22Error, PSP22};
+
+    #[derive(Debug, PartialEq, Eq, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum DexError {
+        Arithmethic,
+        TokenNotInPool(AccountId),
+        TooMuchSlippage,
+        NotEnoughLiquidityOf(AccountId),
+        InsufficientLiquidityShares,
+        PSP22(PSP22Error),
+    }
+
+    impl From<PSP22Error> for DexError {
+        fn from(e: PSP22Error) -> Self {
+            DexError::PSP22(e)
+        }
+    }
+
+    type Event = <SimpleDex as ContractEventBase>::Type;
+
+    #[ink(event)]
+    pub struct Swapped {
+        caller: AccountId,
+        #[ink(topic)]
+        token_in: AccountId,
+        #[ink(topic)]
+        token_out: AccountId,
+        amount_in: Balance,
+        amount_out: Balance,
+    }
+
+    #[ink(storage)]
+    pub struct SimpleDex {
+        pub swap_fee_percentage: u128,
+        pub pool: Vec<AccountId>,
```

Comment:
Better to wrap this into `Lazy` to make the storage type un-packed. It also important to keep the total counter of the pool members because if the `pool` size grows too large, then you might reach the `16KB` decoding limit which will revert the transaction any time you query the vector

Reply:
Good point, generally a large pool makes little sense - I capped it at 3, enough to demonstrate how a DEX operates